### ISSUE TITLE
[feat]: Show currently installed version of service in marketplace

### ISSUE THREAD

**Issue by elvece** [Start9Labs/embassy-os#1497]
### Prerequisites

- [X] I have searched for [existing issues](https://github.com/start9labs/embassy-os/issues) that already suggest this feature, without success.

### Describe the Feature Request

If a service is currently installed, display the version that is installed in the marketplace page of that service in addition to the one that is currently available. 

### Describe the Use Case

It's confusing which version of a service is installed from the marketplace, and what is the most recently released version. Make this display of valuable information more clear.

**Comment by MattDHill** [comment]
I question the usefulness of displaying the currently-installed version of a service in the marketplace listing page. Why?

**Comment by chrisguida** [comment]
That's because Apple and Google auto-update your apps. The user doesn't know or care what version they're on. In embassy-os it's a different story because of all the interdependencies.

**Comment by MattDHill** [comment]
Ok forget about the comparison. Why would someone care about this?

**Comment by kn0wmad** [comment]
I think simply displaying it as "v0.22 -> v0.23" shows both where you are and where you are headed, it gives an instant representation of how big the leap is, which may prompt the user to decide for or against an update, alongside the release notes of course

**Comment by chrisguida** [comment]
....the left hand side _does_ matter, that's why we're asking for this feature. it's annoying to have to look at two different pages to see which version you're on and which one you're going to. you have to memorize the first set of 4 numbers, then find the correct link and click it, then compare the new version you see on your screen with the one you previously memorized. Putting them both on the same page eliminates all of that cognitive overhead. I'm amazed that anyone would not see the benefit to this.

**Comment by chrisguida** [comment]
In addition to the benefits above, this feature would reassure the user that their current version is, indeed, out of date, _and_ helps illustrate just _how_ out of date.

**Comment by chrisguida** [comment]
I agree that this feature would be less useful on a different system, but on our system our users need to decide whether they even want to update. This feature aids them in making that decision.

**Comment by chrisguida** [comment]
>A decision to update or not update seems unrelated to your current version and entirely related to whether or not you want the new version.

If a user doesn't know what version they're on, they might think that they already have the latest version. This feature eliminates that doubt.

**Comment by MattDHill** [comment]
> > A decision to update or not update seems unrelated to your current version and entirely related to whether or not you want the new version.
> 
> The current version is absolutely relevant. If a user doesn't know what version they're on, they might think that they already have the latest version. This feature eliminates that doubt.

Just saw this after sending my previous comment. The giant "Update Available" on the screen should make it clear whether or not they are on the current version.

**Comment by MattDHill** [comment]
If they are on the latest version, it will just say installed beside the latest version which is already displayed.

**Comment by chrisguida** [comment]
"Update available" is too vague. Users should be able to clearly evaluate exactly what changes they are getting so they can make a proper decision about whether to update.

**Comment by chrisguida** [comment]
A user might also opt to update to a version that *isn't* the latest version; this would also aid in that situation.

**Comment by chrisguida** [comment]
Again, it's amazing to me that anyone would not see the benefit to _being able to see what version you're currently on_. What version you're currently on is relevant in all sorts of situations; for instance, when something unexpected happens or when the service you're running is missing a feature you thought it had. Then you can just go look at what version you have, which should explain what's going on.

**Comment by MattDHill** [comment]
"it's amazing to me that anyone would not see the benefit to being able to see what version you're currently on" is just a reference to me and is pretty insulting. But no, I do not see the benefit of being able to see what version I'm currently on *in the marketplace show page*. Nor do I think others will benefit from this. I can see plenty of reasons why someone would want to see what version they are on in other situations, like the ones you mention ("when something unexpected happens or when the service you're running is missing a feature you thought it had"), but neither of these situations are relevant when you have already made the explicit decision to view the latest version of a service. If I want to see what version I have, it is front and center on the installed service show page. Given that the status quo is that we do not display the current version, I'm going to close this ticket unless I hear a reason why someone would want to see the currently installed version of a service on the marketplace show page.

**Comment by chrisguida** [comment]
>but neither of these situations are relevant when you have already made the explicit decision to view the latest version of a service. 

This is not true, for reasons I've already explained. A sovereign user wants to be able to see a clearly delineated diff between the feature sets of the version they're currently using and the version they are looking at in the marketplace. And again, they will not necessarily be looking at the latest version. They may be looking at versions that are between where they are and the latest version. The only way they can understand the difference between where they are and where they want to go is if they can see both versions, side by side. The reason we have a "view in marketplace" button is so the user can see this information. If this information isn't useful, then why not just replace the "view in marketplace" button with an "update to latest" button?

There are tons of pros to this feature and no cons. Or maybe I'm not understanding the cons?

**Comment by MattDHill** [comment]
Ultimately, we disagree here. I understand you would like this, but I remain unconvinced it would be useful for the majority of users. The marketplace show page has been slightly redesigned in 0.3.1 for clarity. Let's see how that plays. We can revisit this later if others agree it might be worthwhile.

**Comment by elvece** [comment]
The original issue here stemmed from the lack of clarity on the header of the marketplace show page. Mainly, it was confusing to me to understand what version I was *updating to*. 

This is certainly a more apparent issue to service developers and beta testers, because they are first and foremost concerned with an awareness of what version of a service they are on, and what they are updating to, to ensure good reporting of issue that might arise from an update. This action of updating is performed *much* more frequently than a regular user. 

As a regular user, I would be concerned with (1) knowing if an update is available (2) what cool new features I get in that update. Power users might be concerned with specific versions, but as a power user/tester/developer, it is expected that you put more effort into gathering the information you need.

We are ultimately not developing a platform for developers. However, I'm of the opinion that we should enable an admin/power/developer user view that can be intentionally  turned on to view more critical specific information throughout the interface. This way, clarity is prioritized for a regular user in a decluttered fashion, while additional details or power actions are displayed if opted in to this view. 

Further, a marketplace experience should highlight what's available and be agnostic to what you currently have installed. It's simply a shopping room with the hottest and latest available.

I like the solution Matt and I designed for now. There is likely a better solution from a full redesign...we'll get there 🙂

**Comment by ProofOfKeags** [comment]
> We are ultimately not developing a platform for developers. However, I'm of the opinion that we should enable an admin/power/developer user view that can be intentionally turned on to view more critical specific information throughout the interface.

Sounds like a feature to bury in developer mode.

**Comment by chrisguida** [comment]
Here is a perfect example of why the current version needs to go on the marketplace page:

Specter tells me I don't have the correct version of bitcoind:
<img width="1440" alt="Screen Shot 2022-06-10 at 2 49 38 PM" src="https://user-images.githubusercontent.com/7445670/173140682-6557c3b9-9fd2-4015-873c-4224dcc65289.png">

So I click "Update" which brings me straight to the marketplace show page for bitcoind:
<img width="1440" alt="Screen Shot 2022-06-10 at 2 53 01 PM" src="https://user-images.githubusercontent.com/7445670/173140971-45503799-f94b-40a3-a2a1-163756c77b33.png">

Nowhere on the page does it indicate what version I _do_ have. This is extremely upsetting, because it feels like I'm being forced to install a particular version of bitcoind, whereas I thought the version of bitcoind I had previously was perfectly adequate. Without knowing what other options there are, I have only one choice. As developers of software for sovereign individuals, we should never trap our users this way.

I don't think this is just a developer thing; I think this is an important feature to the mission of this project.

However, I'll certainly accept an implementation of this hidden behind the developer tools feature, if nothing else.

**Comment by chrisguida** [comment]
What I don't get is why all of the other stuff on this page is not considered clutter, while this very small, very useful piece of information _is_ considered clutter?

**Comment by ProofOfKeags** [comment]
I feel like the main reason to include the version you currently have installed is because we haven't yet automated away 100% of decisionmaking with respect to version management. We provide good recommendations but the end user is not yet free to ignore versions and so resolving these conflicts is something that users will have to do from time to time. As such, providing them with the tools to be able to figure this out is better than not.

I'm somewhat in agreement with Chris here that it feels like any opposition to putting this info on the page seems misplaced.

**Comment by MattDHill** [comment]
Example. I'm trying to decide whether or not to update to 23.0.0. Why does my current version help me make this decision?

**Comment by ProofOfKeags** [comment]
Sometimes people wish to be N versions behind the tip (since sometimes new updates can break things). Knowing how far behind you are can be a good motivator to upgrade things. If your position is that people should always update to the newest stuff all the time (a belief I think is going to be harmful once we are no longer the sole source of packages), then it won't matter. However, if we expect (which again I think we should) people to be *intentionally* behind, then this is a helpful feature.

fwiw this was a feature since the 0.1 days. I remember writing it into the original agent. So removing it is technically a feature reduction, and should be done because it helps the user (which I do not believe it does). I am generally sympathetic to the clutter argument, but I personally don't think that offering the version that is currently installed really constitutes clutter, or at least it doesn't have to.

From a more general analogy perspective. Many people only want to go shopping for things unless they have an idea of what they need, and usually they can't have an idea of what they need without knowing what they have. This can be accomplished by taking an inventory before going to the store (we often do this with things like groceries), but imagine if you could, while you were at the grocery store, know everything in your pantry and fridge without having to go home, this would be a superior experience to having to bring all that info with you before going to the store.

Now, the analogy presented above is slightly different: looking up "what you have" is different than being shown proactively what you have, so perhaps there is room for burying this element in an "on hover" menu or a help text element next to "Update Available" or something. This may be the right compromise between keeping things uncluttered and still having the info available conveniently for those who want it.

**Comment by MattDHill** [comment]
Thanks for the explanation. I retain that the dominant/only consideration when deciding to purchase or upgrade a product, any product, is "What am I getting?" _not_ "What do I have?". The latter is useful only insofar as it clarifies the former. The diff is what matters. My stance is that knowing the current version of a service does not help with this, because the version numbers alone are not meaningful, except to a tiny minority of power users and devs. To the average user `21.0 -> 23.0` and `22.0 -> 23.0` are effectively the same thing. In fact, showing the version diff may actually cause unnecessary cognitive overhead to the user who does not know or care about the numbers. For the person who does care, you probably know your current version already, and if you do not, you can discover it with a single button in the latest UI. The biggest and possibly only real beneficiary of having this information available front and center is a dev who is constantly installing and updating services for testing and loses track of where they are, ie the people advocating for this feature in this thread.

My ask remains the same as above:

"The marketplace show page has been slightly redesigned in 0.3.1 for clarity. Let's see how that plays. We can revisit this later if others agree it might be worthwhile."

**Comment by ProofOfKeags** [comment]
>> The diff is what matters.

> Exactly, and without knowing what version you're on, you cannot determine the diff without extreme effort. The marketplace show page shows only what has happened since the last version, not since the version you have.

Actually this can be accomplished by trimming the release notes to only include the release notes of the diff. I agree that the diff is the only important consideration. And I also agree that version numbers are a suboptimal way to express that to someone who doesn't keep up with what features are in what version. 

> I disagree. The average bitcoin user who is running their own node is keeping track of updates coming from bitcoin core and knows roughly how old each version is, and even some features each is likely to have added. 

This is likely true, however the feature consideration in this ticket is not strictly limited to Bitcoin. As such, we can't really only consider Bitcoin users for it.

> "The marketplace show page has been slightly redesigned in 0.3.1 for clarity. Let's see how that plays. We can revisit this later if others agree it might be worthwhile."

I don't see an issue with this. If there is an easy way from that view to get the current version info even if it isn't displayed initially, I think that is the right compromise.
